thread-manager/src/policy.rs

```diff
@@ -0,0 +1,114 @@
+#[cfg(target_os = "linux")]
+use thread_priority::{NormalThreadSchedulePolicy, ThreadExt, ThreadSchedulePolicy};
+use {
+    serde::{Deserialize, Serialize},
+    std::sync::OnceLock,
+};
+
+#[cfg(not(target_os = "linux"))]
+#[derive(Clone, Copy)]
+pub(crate) struct ThreadSchedulePolicy {}
+
+static CORE_COUNT: OnceLock<usize> = OnceLock::new();
+
+pub const DEFAULT_PRIORITY: u8 = 0;
+
+#[derive(Default, Debug, Clone, Serialize, Deserialize)]
+pub enum CoreAllocation {
+    ///Use OS default allocation (i.e. do not alter core affinity)
+    #[default]
+    OsDefault,
+    ///Pin each thread to a core in given range. Number of cores should be >= number of threads
+    PinnedCores { min: usize, max: usize },
+    ///Pin the threads to a set of cores
+    DedicatedCoreSet { min: usize, max: usize },
+}
+
+impl CoreAllocation {
+    /// Converts into a vector of core IDs. OsDefault is converted to empty vector.
+    pub fn as_core_mask_vector(&self) -> Vec<usize> {
+        let core_count = CORE_COUNT.get_or_init(num_cpus::get);
+        match *self {
+            CoreAllocation::PinnedCores { min, max } => (min..max).collect(),
+            CoreAllocation::DedicatedCoreSet { min, max } => (min..max).collect(),
+            CoreAllocation::OsDefault => Vec::from_iter(0..*core_count),
+        }
+    }
+}
+
+#[cfg(target_os = "linux")]
+pub fn set_thread_affinity(cores: &[usize]) {
+    assert!(
+        !cores.is_empty(),
+        "Can not call setaffinity with empty cores mask"
+    );
+    if let Err(e) = affinity::set_thread_affinity(cores) {
+        let thread = std::thread::current();
+        let msg = format!(
+            "Can not set core affinity {:?} for thread {:?} named {:?}, error {e}",
+            cores,
+            thread.id(),
+            thread.name()
+        );
+        panic!("{}", msg);
```

Comment:
We don't need to `format` then call `panic` which will `format` internally the string into your "{}"
```suggestion
        panic!(
            "Can not set core affinity {:?} for thread {:?} named {:?}, error {}",
            cores,
            thread.id(),
            thread.name(),
            e
        );
```